### ISSUE TITLE
Build up of containers eventually chokes system resources

### ISSUE THREAD

**Issue by levithomason** [TechnologyAdvice/DevLab#54]
After running several `lab` tasks throughout the day, the containers started by DevLab build up over time.  These eventually turn my fans into rocket ships and start choking my system:

<p align=center>
  <img width=550 src=https://cloud.githubusercontent.com/assets/5067638/24432876/c8c445d4-13d9-11e7-86ad-1d8cc6bf0fdf.png />
</p>

I find I have to restart docker periodically to reclaim system performance and settle my fans.  It would be grand if we could find a way to reliably kill off these containers after using them.

If we know they are running, can we spawn a child process to kill them instead of logging that they are still running?

**Comment by levithomason** [comment]
To help illustrate the prevalence of the issue, these are now the containers that have accumulated in the 15m since I posted the issue:

<p align=center>
  <img width=550 src=https://cloud.githubusercontent.com/assets/5067638/24433383/0e2bbf6e-13dd-11e7-9ffe-4e7966dc23cc.png />
</p>

**Comment by Fluidbyte** [comment]
The previous versions of Devlab did not do a "headless" shutdown. After the task completed the `stop` and `rm` tasks were run before the parent process exited. We moved this to a forked process because with multiple services this shutdown process can take in upwards of 10s with some of the containers we were using.

I'm more than happy to restore the original in-process shutdown of service containers, but would like to get some other thoughts before pulling the trigger.

**Comment by Fluidbyte** [comment]
cc: @TomFrost

**Comment by levithomason** [comment]
Is it possible to also kick off `lab --cleanup` in another process if running containers are found?  In contrast to only listing them and giving the user a command to clean them up.

```
⚠ These containers did not exit: ...
⚠ You can remove them with `lab --cleanup`, we'll try that now.
```

This would just kick off `lab --cleanup` along with the task that was executed.

**Comment by Fluidbyte** [comment]
After thinking about this for a couple days I feel like the best option is to return to in-process shutdown of service containers. It adds a few seconds but was very consistent previously.

We _could_ go with your suggestion above, however, it's really just deferring the issue and it seems like detached processes under heavy load (which is what we're seeing) is hit or miss.

**Comment by levithomason** [comment]
Might be the best option.  My worry with it is the same as the original issue, how long it will take. 

Timing it:
- 2 stray containers `lab --cleanup` 21s
- 20 stray containers `--lab cleanup` 1m 42s

On this particular project properly pronounced with a "p", there are 5 service containers.  With it taking 5-10s per container to clean up and ~5s to start them, it seems there would be ~30s - 1m of overhead time involved in starting/stopping tasks.

If this is the best option, and truly the only option, I think it is better to leave the resource hogs for now.  My reasoning is that I'd rather stop for a couple minutes to restart docker when it is convenient for me than add a full minute to every task run.

Thoughts?

**Comment by Fluidbyte** [comment]
I think the addressing the [serial shutdown process](https://github.com/TechnologyAdvice/DevLab/blob/master/src/services.js#L72-L73) could be a good way to address this.

The approach previously has been to join `stop` and `rm` commands with `&&`, then chain those. I've tried manual startup and shutdowns on numerous containers and even on some of the "heavier" containers, run alone, the shutdown time is 3-4s.

I'm thinking if I reworked the shutdown script to run these in parallel we could 1. stop all service containers quickly and 2. be able to report failures immediately, allowing users to see them before running another process.

**Comment by Fluidbyte** [comment]
To preemptively address the follow-up question of why _not_ continue to do this in a forked process?

The consistency here seems to be an increase in relics when the host system is under load. I don't see any reason to have the system create additional "threads" if I can prove spawned processes under the parent process to be efficient. If this proves itself I'll go this route, if there's a dramatic increase in time (upwards of 10s) then it may be prudent to go the route of headless forks.

Considering that services not shutting down only contributes to additional load, and thus would mean an increased chance of future services not shutting down, it seem prudent to nip this in the bud during the running process and give the user an indication of an issue before running another task.

**Comment by TomFrost** [comment]
I think this is all very sane-- I'm still curious why this is happening at all, though. It happened with Stephen yesterday as well, and I've noticed that both he and Levi use Ctrl+C to kill the lab script _a lot_ when I pair with them, often times while the supporting containers are still starting up. Perhaps there's a window during that startup in which they can send SIGINT without triggering the shutdown process?

**Comment by Fluidbyte** [comment]
I thought it might be that, but even running a few back-to-back locally on my Mac I got the same thing (just running `env` tasks). I agree early kills should during service start should be handled, but can confirm it's not the cause.

I got the parallel shutdown process mocked out and it's fast. Does add a couple seconds but not bad on most scenarios, about as fast as (if not faster than) the startup. The best news is that it has prevented any failures in my tests thus far.

I did discover one thing though - a lot of time-suck of shutdown could be circumvented by finding better/faster containers. The Elasticsearch image on one particularly problematic project was a big pain and (alone) takes about 12-13s to shut down just using standalone `docker stop`.
